lib/translator/write.cc

```diff
@@ -0,0 +1,242 @@
+                            transfer_length_pages, buffer_out);
+
+  if (status_code != StatusCode::kSuccess) {
+    return status_code;
+  }
+
+  // cdw10 Starting lba bits 31:00
+  uint32_t host_endian_lba = (write_cmd.logical_block_address_1 << 16) +
+                             ntohs(write_cmd.logical_block_address_2);
+
+  nvme_wrapper.cmd.cdw[0] = htoll(host_endian_lba);
+  nvme_wrapper.cmd.cdw[2] =
+      htoll(static_cast<uint32_t>(updated_transfer_length - 1));
+  return status_code;
+}
+
+StatusCode Write10ToNvme(Span<const uint8_t> scsi_cmd,
+                         NvmeCmdWrapper& nvme_wrapper, Allocation& allocation,
+                         uint32_t nsid, uint32_t page_size, uint32_t lba_size,
+                         Span<const uint8_t> buffer_out) {
+  scsi::Write10Command write_cmd = {};
+  if (!ReadValue(scsi_cmd, write_cmd)) {
+    DebugLog("Malformed Write10 Command");
+    return StatusCode::kInvalidInput;
+  }
+
+  StatusCode status_code = Write(
+      write_cmd.fua, write_cmd.wr_protect, ntohs(write_cmd.transfer_length),
+      nvme_wrapper, allocation, nsid, page_size, lba_size, buffer_out);
+
+  if (status_code != StatusCode::kSuccess) {
+    return status_code;
+  }
+
+  nvme_wrapper.cmd.cdw[0] = htonl(write_cmd.logical_block_address);
+  return status_code;
+}
+StatusCode Write12ToNvme(Span<const uint8_t> scsi_cmd,
+                         NvmeCmdWrapper& nvme_wrapper, Allocation& allocation,
+                         uint32_t nsid, uint32_t page_size, uint32_t lba_size,
+                         Span<const uint8_t> buffer_out) {
+  scsi::Write12Command write_cmd = {};
+  if (!ReadValue(scsi_cmd, write_cmd)) {
+    DebugLog("Malformed Write12 Command");
+    return StatusCode::kInvalidInput;
+  }
+
+  StatusCode status_code = Write(
+      write_cmd.fua, write_cmd.wr_protect, ntohl(write_cmd.transfer_length),
+      nvme_wrapper, allocation, nsid, page_size, lba_size, buffer_out);
+
+  if (status_code != StatusCode::kSuccess) {
+    return status_code;
+  }
+
+  nvme_wrapper.cmd.cdw[0] = htonl(write_cmd.logical_block_address);
+  return status_code;
+}
+
+StatusCode Write16ToNvme(Span<const uint8_t> scsi_cmd,
+                         NvmeCmdWrapper& nvme_wrapper, Allocation& allocation,
+                         uint32_t nsid, uint32_t page_size, uint32_t lba_size,
+                         Span<const uint8_t> buffer_out) {
+  scsi::Write16Command write_cmd = {};
+  if (!ReadValue(scsi_cmd, write_cmd)) {
+    DebugLog("Malformed Write16 Command");
+    return StatusCode::kInvalidInput;
+  }
+
+  StatusCode status_code = Write(
+      write_cmd.fua, write_cmd.wr_protect, ntohl(write_cmd.transfer_length),
+      nvme_wrapper, allocation, nsid, page_size, lba_size, buffer_out);
+
+  if (status_code != StatusCode::kSuccess) {
+    return status_code;
+  }
+
+  nvme_wrapper.cmd.cdw[0] = htoll(write_cmd.logical_block_address);
+  nvme_wrapper.cmd.cdw[1] = htoll(write_cmd.logical_block_address >> 32);
```

Comment:
This should follow the format of `read.cc` https://github.com/googleinterns/scsi2nvme/blob/master/lib/translator/read.cc#L235 where convert the value to host endian (`ntohll`) and then do the static_cast or bit shift, and then call `htoll`